pkg/controller/job/job_controller.go

```diff
@@ -1150,6 +1162,17 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
 			jobCtx.finishedCondition = newFailedConditionForFailureTarget(jobCtx.finishedCondition, jm.clock.Now())
 		}
 	}
+	if isSuccessCriteriaMetCondition(jobCtx.finishedCondition) {
+		// Append the interim SuccessCriteriaMet condition to update the job status with before finalizers are removed.
+		if successCriteriaMetCondition(jobCtx.job) == nil {
```

Comment:
What is the reason we do this `if` but we don't in FailureTarget? Maybe we should have added it there too, but I never found a scenario in which this is needed. Anyone knows?

Reply:
Oh, I see, because for `FailureTarget` we transform it into `Failed` condition early on: https://github.com/kubernetes/kubernetes/blob/24123d1972d71b2297c6efb6b7bb55e46071f809/pkg/controller/job/job_controller.go#L847. So we only observe `FailureTarget` condition on the first sync. For `SuccessCriteriaMet` we don't do that. I'm wondering which pattern is better, but least I see why the difference.

Reply:
I'd probably suggest naming these methods a bit better. I will admit their current names are a bit confusing, and I always end up looking at what they do :sweat:

Reply:
@soltysh Based on Mihal's proposal, I will update `successCriteriaMetCondition` with `hasSuccessCriteriaMetCondition `.

https://github.com/kubernetes/kubernetes/pull/123412/files#r1516071148